server/api/src/rsDataObjOpen.cpp

```diff
@@ -185,354 +197,343 @@ namespace
-            replica.replica_status(INTERMEDIATE_REPLICA);
+                    return l1_index;
+                }
 
-            l1_index = populate_L1desc_with_inp(dataObjInp, replica, dataObjInp.dataSize);
+                irods::log(LOG_ERROR, fmt::format(
+                    "[{}:{}] - L1 descriptor out of range "
+                    "[fd=[{}], path=[{}], hierarchy=[{}]",
+                    __FUNCTION__, __LINE__, l1_index, _inp.objPath, hierarchy));
 
-            L1desc[l1_index].l3descInx = create_physical_file(*rsComm, l1_index);
+                return SYS_FILE_DESC_OUT_OF_RANGE;
+            }
 
-            return l1_index;
-        }
-        catch (const irods::exception& e) {
-            irods::log(LOG_NOTICE, fmt::format(
-                "{}: create_physical_file of {} failed, status = {}",
-                __FUNCTION__, L1desc[l1_index].dataObjInfo->filePath, e.code()));
-            freeL1desc(l1_index);
-            return e.code();
-        }
-    } // specCollSubCreate
+            auto& l1desc = L1desc[l1_index];
 
-    auto create_new_replica(rsComm_t& _comm, dataObjInp_t& _inp, DataObjInfo* _existing_replica_list) -> int
-    {
-        auto cond_input = ix::make_key_value_proxy(_inp.condInput);
+            if (const int ec = getFilePathName(&_comm, new_replica.get(), l1desc.dataObjInp); ec < 0) {
+                freeL1desc(l1_index);
 
-        rodsObjStat_t* stat{};
-        const irods::at_scope_exit free_obj_stat_out{
-            [&stat]() {
-                freeRodsObjStat(stat);
+                irods::log(LOG_ERROR, fmt::format(
+                    "[{}] - failed to get file path name "
+                    "[error_code=[{}], path=[{}], hierarchy=[{}]",
+                    __FUNCTION__, __LINE__, ec, _inp.objPath, hierarchy));
+
+                return ec;
             }
-        };
 
-        cond_input[SEL_OBJ_TYPE_KW] = "dataObj";
-        if (const int ec = rsObjStat(&_comm, &_inp, &stat); ec >= 0) {
-            if (stat) {
-                if (COLL_OBJ_T == stat->objType) {
-                    return USER_INPUT_PATH_ERR;
-                }
+            auto l1_cond_input = irods::experimental::make_key_value_proxy(l1desc.dataObjInp->condInput);
+            l1_cond_input[REGISTER_AS_INTERMEDIATE_KW] = "";
+            l1_cond_input[FILE_PATH_KW] = new_replica.physical_path();
+            l1_cond_input[DATA_SIZE_KW] = std::to_string(0);
 
-                if (stat->specColl) {
-                    // Linked collection should have been translated by this point
-                    if (LINKED_COLL == stat->specColl->collClass) {
-                        return SYS_COLL_LINK_PATH_ERR;
-                    }
-
-                    if (UNKNOWN_OBJ_T == stat->objType) {
-                        return specCollSubCreate(&_comm, _inp );
-                    }
+            if (!brand_new_data_object) {
+                // TODO: How should we lock the data object before creating?
```

Comment:
When a new replica is created, the data object is not locked until after the replica has been registered in the catalog. For a brand new data object, this is not a problem because there are no replicas to lock, but for an existing data object there is a gap between creating the new replica and locking the other replicas. Since ILL operates on the RST, we will need to insert the RST entry for the existing data object (if there is one), lock the data object, then insert the entry for the new replica. 

But I never really sat down and evaluated whether that was the best approach. Open to suggestions.

Reply:
so, the window of opportunity for shenanigans is limited to the time a new replica is registered (and this itself is limited to being a new replication operation, not new incoming data b/c no new replicas with force flag) and the lock applied to the siblings...

and i think the only shenanigans that could happen is an identical replication event to a different separate resource...

so the question is... what happens in that case?   they both are allowed to register a new replica, one of them gets the lock, the other is halted/errors, and ... you end up with a single brand new stale replica, and a single brand new good replica.   i think.

Reply:
Right, if the same replica information is used, a duplicate entry error will occur in the database when we go to register it. So I think what you've said is right. I believe the scenario describes something similar to what is happening in #4313 and #3930. That is, a replication to random resource hierarchy could choose two different hierarchies given concurrent uploads and successfully create both replicas.

I'm not 100% sure what would happen in this case, but my guess is that the first operation would complete and unlock the data object. The second object, unawares, would then complete its operation and do another unlock. If it attempts to restore the replica statuses of the sibling replicas, there will be some failures because the first operation should have removed the values from the `data_status` column (in the future, there will be a list of PIDs and hostnames which may prevent such errors). It may result in an extra, stale replica, as you said.

This is better than an extra replica marked good, but is also a hole in the defense provided by logical locking. Should we document the issue and move on for now, or should we try to fix this now?

Reply:
I created #5542 to address this issue. The system will not leave any data objects locked as all locks and unlocks in the catalog are atomic and only occur once. As such, the current implementation does not present any major issues in my view.

Reply:
agreed, thanks.